sources/kube.go

```diff
@@ -66,11 +68,39 @@ func (self *KubeSource) recordNodeError(name string) {
 	self.nodeErrors[name]++
 }
 
-func (self *KubeSource) recordPodError(podName, podId, hostIp string) {
+func (self *KubeSource) addActivePodInstance(podName, podId, hostIp string) {
+	podInstance := PodInstance{Pod: podName, PodId: podId, HostIp: hostIp}
+	self.activePods[podInstance] = true
+}
+
+func (self *KubeSource) cleanupPods(pods []Pod) {
 	self.stateLock.Lock()
 	defer self.stateLock.Unlock()
+	newActivePods := make(map[PodInstance]bool)
+	newPodErrors := make(map[PodInstance]int)
+	for _, pod := range pods {
+		podInstance := PodInstance{Pod: pod.Name, PodId: pod.ID, HostIp: pod.HostIP}
+		if self.activePods[podInstance] {
+			newActivePods[podInstance] = true
+			if self.podErrors[podInstance] != 0 {
+				newPodErrors[podInstance] = self.podErrors[podInstance]
+			}
+		}
+	}
+	self.activePods = newActivePods
+	self.podErrors = newPodErrors
+}
 
+func (self *KubeSource) recordPodError(podName, podId, hostIp string) {
+	self.stateLock.Lock()
+	defer self.stateLock.Unlock()
 	podInstance := PodInstance{Pod: podName, PodId: podId, HostIp: hostIp}
+
+	// Heapster knows about pods before they are up and running on a node.
+	// Ignore errors till we have seen an instance up at least once.
```

Comment:
It just occurred to me that another approach to handle this is to ignore Pods that are not in "Running" state.